### PR TITLE
make dns record ttl configurable

### PR DESCRIPTION
**Pull request by ChenglongGao327**
If dns capture is enabled, different scenarios have different requirements for dns recording ttl, So making this configurable.

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pkg/istio-agent/agent.go [hunk 1]
```diff
@@ -467,7 +467,7 @@ func (a *Agent) Run(ctx context.Context) (func(), error) {
 }
 
 func (a *Agent) initLocalDNSServer() (err error) {
-	// we dont need dns server on gateways
+	// we don't need dns server on gateways
```
Comment (howardjohn) [inline pkg/istio-agent/agent.go:470]: Please avoid refactoring in a PR with real changes.. it makes it hard to review

**Review by howardjohn — CHANGES_REQUESTED** [review]
Can you discuss the use case a bit more? I think we want to ensure there are good reasons to configure things before we add more hacky env vars

**Comment by ChenglongGao327** [comment]
> Can you discuss the use case a bit more? I think we want to ensure there are good reasons to configure things before we add more hacky env vars

For scenarios that run stably and require frequent domain name resolution, set the ttl to a larger value to reduce the number of times to access DNS; in other scenarios, ttl can be set to be more general.

File: pkg/istio-agent/agent.go (hunk 1 above)
Comment (ChenglongGao327) [reply pkg/istio-agent/agent.go:470]: > Please avoid refactoring in a PR with real changes.. it makes it hard to review

Sorry,  i got it

**Comment by howardjohn** [comment]
is it something that makes sense as a global option or per service? I
cannot imagine needing low TTL for a cluster ip - it literally never
changes. But for headless service smaller TTL makes more sense

**Comment by ChenglongGao327** [comment]
> is it something that makes sense as a global option or per service? I cannot imagine needing low TTL for a cluster ip - it literally never changes. But for headless service smaller TTL makes more sense

Yes, the clusterIP ttl needs to be set larger, and the headless ttl needs to be set smaller. 
I want to distinguish headless service and clusterIP，and set different ttl. such as through kubernetes and kube-dns clusterIP
When connecting to k8s, if the IP address corresponding to the service domain name is not in the same network segment as the kubernetes and kube-dns service cluster IP addresses, then the service is considered a headless service.

**Comment by howardjohn** [comment]
Do you have a real use case for this, or just saw a TODO? I would avoid fixing TODOs without a strong motivation, usually they mean "TODO [if we ever have a compelling use case])".

> IP address corresponding to the service domain name is not in the same network segment as the kubernetes
We do not need this implicit logic, we know if its headless or not in the control plane

**Comment by ChenglongGao327** [comment]
> Do you have a real use case for this, or just saw a TODO? I would avoid fixing TODOs without a strong motivation, usually they mean "TODO [if we ever have a compelling use case])".

I saw a TODO and think that it is better to set different ttl for different scenarios, so making it configurable

**Comment by howardjohn** [comment]
I highly recommend not implementing TODOs like this - the reason they were
not done is that they add a lot of complexity and we haven't seen any real
world demand to warrant that complexity

**Comment by ChenglongGao327** [comment]
> I highly recommend not implementing TODOs like this - the reason they were not done is that they add a lot of complexity and we haven't seen any real world demand to warrant that complexity

ok, i see. Thanks for your reviews

**Pull request closed without merging**